provider/data_source_keycloak_realm_keys.go

```diff
@@ -0,0 +1,189 @@
+package provider
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/mrparkers/terraform-provider-keycloak/keycloak"
+)
+
+func dataSourceKeycloakRealmKeys() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceKeycloakRealmKeysRead,
+		Schema: map[string]*schema.Schema{
+			"realm_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"algorithms": {
+				Type:     schema.TypeSet,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Optional: true,
+			},
+			"status": {
+				Type:     schema.TypeSet,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Optional: true,
+			},
+			"keys": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"algorithm": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+						"certificate": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+						"provider_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+						"provider_priority": {
+							Type:     schema.TypeInt,
+							Computed: true,
+							Optional: true,
+						},
+						"kid": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+						"public_key": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+						"status": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+						"type": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func flattenRealmKeys(realmKeys []keycloak.Key) []map[string]interface{} {
+	keyMap := make([]map[string]interface{}, 0)
+	for _, key := range realmKeys {
+		element := make(map[string]interface{})
+		if key.Algorithm != nil {
+			element["algorithm"] = key.Algorithm
+		}
+		if key.Certificate != nil {
+			element["certificate"] = key.Certificate
+		}
+		if key.ProviderId != nil {
+			element["provider_id"] = key.ProviderId
+		}
+		if key.ProviderPriority != nil {
+			element["provider_priority"] = key.ProviderPriority
+		}
+		if key.Kid != nil {
+			element["kid"] = key.Kid
+		}
+		if key.PublicKey != nil {
+			element["public_key"] = key.PublicKey
+		}
+		if key.Status != nil {
+			element["status"] = key.Status
+		}
+		if key.Type != nil {
+			element["type"] = key.Type
+		}
+
+		keyMap = append(keyMap, element)
+	}
+	return keyMap
+}
+
+func setRealmKeysData(data *schema.ResourceData, keys *keycloak.Keys) error {
+	data.SetId(data.Get("realm_id").(string))
+
+	err := data.Set("keys", flattenRealmKeys(keys.Keys))
+	if err != nil {
+		return fmt.Errorf("could set keys with values '%+v'\n%+v", keys.Keys, err)
```

Comment:
this error message is a bit unclear to me

Reply:
I fixed the error message. Normally `data.Set()`should not fail, but when the data (i.e result of `flattenRealmKeys()`) has the wrong structure than an error might occur.